sdk/attestation/Azure.Security.Attestation/src/AttestationToken.cs

```diff
@@ -143,43 +202,390 @@ protected AttestationToken()
+            }
+
+            if (!ValidateTokenSignature(possibleCertificates))
+            {
+                throw new Exception($"Could not find a certificate which was used to sign the token.");
+            }
+
+            if (!ValidateCommonProperties(options))
+            {
+                return false;
+            }
+            if (options.ValidationCallback != null)
+            {
+                return options.ValidationCallback(this, SigningCertificate);
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Returns a set of candidate signing certificates based on the contents of the token.
+        /// <para/>
+        /// If the caller provides a set of signers, than that set of signers will be used as the complete set of candidates for signing.
+        /// If the caller does not provide a set of signers, then the <see cref="GetCandidateSigningCertificates(IReadOnlyList{AttestationSigner}, CancellationToken)"/>
+        /// API will a set of callers derived from the contents of the attestation token.
+        /// </summary>
+        /// <param name="attestationSigners">The desired set of signers for this token - if present, then this is the exclusive set of signers for the token.</param>
+        /// <param name="cancellationToken">Cancellation token used to cancel this operation.</param>
+        /// <returns>The set of <see cref="AttestationSigner"/> which might be used to sign the attestation token.</returns>
+        private Task<AttestationSigner[]> GetCandidateSigningCertificatesAsync(IReadOnlyList<AttestationSigner> attestationSigners, CancellationToken cancellationToken = default)
+        {
+            return Task.FromResult(GetCandidateSigningCertificates(attestationSigners, cancellationToken));
+        }
+
+        /// <summary>
+        /// Returns a set of candidate signing certificates based on the contents of the token.
+        /// </summary>
+        /// <param name="attestationSigners">The desired set of signers for this token - if present, then this is the exclusive set of signers for the token.</param>
+        /// <param name="cancellationToken">Cancellation token used to cancel this operation.</param>
+        /// <returns>The set of <see cref="AttestationSigner"/> which might be used to sign the attestation token.</returns>
+        private AttestationSigner[] GetCandidateSigningCertificates(IReadOnlyList<AttestationSigner> attestationSigners, CancellationToken cancellationToken = default)
         {
-            Argument.AssertNotNull(attestationSigningCertificates, nameof(attestationSigningCertificates));
-            Argument.AssertNotNull(attestationSigningCertificates[0], nameof(attestationSigningCertificates));
-            Argument.AssertNotNullOrEmpty(attestationSigningCertificates, nameof(attestationSigningCertificates));
+            string desiredKeyId = Header.KeyId;
+
+            if (cancellationToken.IsCancellationRequested)
```

Comment:
Q: How much value is this cancellation token if only checked once?

Reply:
None. But the pattern is that public APIs that can be asynchronous come in two forms: A synchronous api and an async API, both of which take cancellation tokens. The only reason the token is checked here is to avoid a compiler diagnostic.

At some point, this code will likely have logic to retrieve values off-node (JKU/XCU support). At that point, the async functionality will become critical.